pkg/virt-operator/kubevirt.go

```diff
@@ -444,7 +445,7 @@ func (c *KubeVirtController) genericAddHandler(obj interface{}, expecter *contro
 		if expecter != nil {
 			expecter.CreationObserved(controllerKey)
 		}
-		c.queue.Add(controllerKey)
+		c.queue.AddAfter(controllerKey, defaultAddDelay)
 	}
```

Comment:
i'm not sure this gives the desired result. From what I remembered, we wanted to ensure that the reconcile loop only pops every 'x' seconds. By adding this delay, what behavior are we actually getting? is this simply shifting the queue popping 5 seconds out where we'll still potentially have lots of reconcile pops within a short period, but just delayed a bit?

It's unclear to me what happens here.

Reply:
I made a mistake in the original PR.. I thought that by creating a custom workqueue limiter(as seen below) any `.Add(key)` function call would respect that workqueue's rate limiting rules. 

The "tldr" is I don't think `Add()` respects any kind of delay or rate limiting. 

```go
rl := workqueue.NewMaxOfRateLimiter(
    workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 1000*time.Second),
    &workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Every(5*time.Second), 1)},
)
```

Then using that for the queue here:
```go 
...
queue:            workqueue.NewRateLimitingQueue(rl),
....
```

However, while trying to debug reconciliation I noticed it was not waiting the anticipated 5 seconds between syncs. 

`NewRateLimitingQueue(rl)` returns the [RateLimitingInterface](https://github.com/kubernetes/client-go/blob/master/util/workqueue/rate_limiting_queue.go#L35-L42)

This interface embeds the `[DelayingInterface](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L30-L34 )` which then embeds `Interface` (which is the top level queue interface). 

Then looking at the `[Add(key)](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L30-L34 )` function it is not aware of the `RateLimitingInterface` and immediately adds the key to the queue to be processed. 

By using `AddAfter(key)` now the key can be processed at quickest every 5 seconds. If instead we use `AddRateLimited` then the delay gets exponentially bigger with each call, which is not desired except for the error case.

Reply:
@ashleyschuett 

Let me make sure I have this straight. If i did something like this.

```
for {
    AddAfter(key)
    sleep(1*time.Second)
}
```

5 Seconds after that loop has started, would be the outcome?

After looking at the AddAfter() code, i don't think the reconcile loop would ever pop. So if we somehow continually get updates for a resource at a cadence < 5 seconds, our operator loop would never get called. 

^ I don't know how likely that is. I just have a feeling that we're using this AddAfter functionality in a way it wasn't intended though.

Reply:
Ahh sorry, now I understand what you are asking. 

The outcome of ^ would be that the reconcile loop would pop every 5 seconds, or whatever amount you pass into the AddAfter function. Due to https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L267 

> ^ I don't know how likely that is. I just have a feeling that we're using this AddAfter functionality in a way it wasn't intended though.

I agree, I don't think this was the intended use case. But I don't think it's expected to throttle the speed of the reconciliation loop for anything other than errors :/

Reply:
> Due to https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L267

oh interesting, so addAfter only guarantees that it will be queued "no later than after x delay", but it could be sooner. 

I guess this will work fine then. I can't really think of any reason why it wouldn't.

Reply:
Yeah the naming is a bit strange for that to be the functionality. My guess is it was done to eliminate the case that you described above.